### ISSUE TITLE
[P1 validation] Dual-MI210 2b parity and 2a throughput

### ISSUE THREAD

**Issue by ewindisch** [hyprstream/hyprstream#317]
Part of #310. Plan: `docs/plans/2026-06-18-multi-gpu-multi-host-inference-spike.md`.

35 GB fits one 64 GB MI210, so split output can be checked against single-GPU output. Add a
parity test (logits/tokens within tolerance) + a 2a replication path with sticky-session routing
groundwork. Acceptance: 2b on 2×MI210 matches single-GPU within numerical tolerance.

**Comment by ewindisch** [comment]
**Retargeted as the post-merge multi-GPU hardware verification gate (2026-06-25).**

The #310 epic is being merged to main with multi-GPU validated in LOGIC + CPU + single-GPU form only:
- #314 split logic: 29/29 CPU split-vs-whole equivalence tests
- #405 device-placement fix (embed/lm_head/norm to device) — merged
- #416 engine-wiring (DevicePool→LayerDeviceMap through TorchEngine) — merged
- #421-E6: real GPU inference on RTX 5090 (single-GPU) + HRW placement + device-pool selection (CPU)

**NOT yet validated: cross-device pipeline-parallel inference on real ≥2-GPU hardware** (the 2×MI210 box). This ticket is that gate: run `HYPRSTREAM_GPU_DEVICES=0,1` on the 2×MI210 machine, confirm the model splits across both GPUs, and verify split output matches single-GPU ground truth within tolerance (1e-3 BF16). This is the epic's core deliverable proven on real multi-GPU hardware — deliberately deferred past the main-merge (hardware is on a separate physical machine), tracked here as the follow-up.

**Comment by ewindisch** [comment]
**Rescoped (Epic B): validation-only.** Remove "add a 2a replication path" from implementation scope — that is now #1235. Retitle intent: **Dual-MI210 2b parity & 2a throughput**.

2b: single-GPU vs split-device comparison of full prefill logits, every cached-decode step, greedy tokens, and errors under declared BF16/FP16 tolerances. Evidence the boundary ran; record peak HBM/card, load time, TTFT, decode tok/s, boundary-copy latency, utilization, MI210 P2P topology/bandwidth; repeat startup/shutdown. Use a factory-supported arch — **not** Qwen3.5 until its factory fallback is removed.

2a: throughput/stickiness/isolation/reassignment/teardown — only after #1235 lands. Depends on reopened #314 + #311. — Epic B ([roadmap](https://claude.ai/code/artifact/8dad534b-c405-4d24-9637-f1781591bab4))

**Comment by ewindisch** [comment]
## Hardware diagnostic result — dual-MI210, gfx90a (2026-07-23)

Ran the current `main` (`69eda149`) inference path on **2× AMD Instinct MI210** with a supported dense Llama checkpoint (SmolLM2-360M, 32 layers).

**Placement works:** even split maps layers 0–15 → `cuda:0`, 16–31 → `cuda:1`; both GPUs hold real layers; clean shutdown.

**Inference fails at the device boundary:**
```
Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cuda:1!
```
PyTorch reports the failing op as `at::_ops::mul_Tensor` (HIP backend) — a device-0 hidden activation reaching the first device-1 decoder layer. The service thread panics; no tokens produced.

This is the **#314 production-2b wiring defect confirmed on real hardware** — not a scale limitation, a wiring bug. Root cause and fix location are in #314.

**Owner:** the `h0fleet` MI210 lane is now implementing the fix on this hardware and will validate to numerical parity before this validation ticket can close. Logs: `~/h0-317-diagnostic.log`, `~/h0-317-gpu-metrics.log`.
